libs/Theseus/src/imex/imex.jl

```diff
@@ -0,0 +1,256 @@
+    # Start actual solve
+    return solve!(integrator)
+end
+
+function solve!(integrator::SimpleImplicitExplicit)
+    @unpack prob = integrator.sol
+
+    integrator.finalstep = false
+
+    while !integrator.finalstep
+        step!(integrator)
+    end # "main loop" timer
+
+    finalize_callbacks(integrator)
+
+    return TimeIntegratorSolution(
+        (first(prob.tspan), integrator.t),
+        (prob.u0, integrator.u),
+        integrator.sol.prob,
+    )
+end
+
+
+function step!(integrator::SimpleImplicitExplicit)
+    @unpack prob = integrator.sol
+    @unpack alg = integrator
+    t_end = last(prob.tspan)
+    callbacks = integrator.opts.callback
+
+    @assert !integrator.finalstep
+    if isnan(integrator.dt)
+        error("time step size `dt` is NaN")
+    end
+
+    # if the next iteration would push the simulation beyond the end time, set dt accordingly
+    if integrator.t + integrator.dt > t_end ||
+            isapprox(integrator.t + integrator.dt, t_end)
+        integrator.dt = t_end - integrator.t
+        terminate!(integrator)
+    end
+
+    # one time step
+    integrator.u_tmp .= integrator.u
+
+    stage!(integrator, alg)
+
+    integrator.u .= integrator.u_tmp
+
+    integrator.iter += 1
+    integrator.t += integrator.dt
+
+    begin
+        # handle callbacks
+        if callbacks isa CallbackSet
+            foreach(callbacks.discrete_callbacks) do cb
+                if cb.condition(integrator.u, integrator.t, integrator)
+                    cb.affect!(integrator)
+                end
+                return nothing
+            end
+        end
+    end
+
+    # respect maximum number of iterations
+    return if integrator.iter >= integrator.opts.maxiters && !integrator.finalstep
+        @warn "Interrupted. Larger maxiters is needed."
+        terminate!(integrator)
+    end
+end
+
+
+function stage!(integrator, alg::RKIMEX)
+    for stage in 1:stages(alg)
+        F! = nonlinear_problem(alg, integrator.f2)
+        # TODO: Pass in `stages[1:(stage-1)]` or full tuple?
+        _, stats = newton_krylov!(
+            F!, integrator.u_tmp, (integrator.u, integrator.dt, integrator.f1, integrator.du, integrator.du_tmp, integrator.p, integrator.t, integrator.stages, integrator.stages_im, stage, integrator.RK), integrator.res;
+            verbose = integrator.opts.verbose, krylov_kwargs = integrator.opts.krylov_kwargs,
+            algo = integrator.opts.algo, tol_abs = 6.0e-6,
+        )
+        @assert stats.solved
+        # Store the solution for each stage in stages
+        ## For a split Problem we need to compute rhs_conservative and rhs_parabolic
+        integrator.f2(integrator.du, integrator.u_tmp, integrator.p, integrator.t + integrator.RK.c_ex[stage] * integrator.dt)
+        integrator.stages[stage] .= integrator.du
+        integrator.f1(integrator.du, integrator.u_tmp, integrator.p, integrator.t + integrator.RK.c_im[stage] * integrator.dt)
+        integrator.stages_im[stage] .= integrator.du
```

Comment:
This part may be problematic since the potentially large Lipschitz constant of the stiff implicit part will affect the accuracy obtained from the implicit solve. Instead, we can use the fact that the Butcher matrix A is (likely or at least in big parts) invertible to get the stage derivative from the numerical solution. See for example https://github.com/ranocha/2025_nls/blob/7fa7880980087c37dcda06cb498395a08aed1174/code/code.jl#L556-L571, where I implemented this for a IMEX methods assuming a linear stiff part.

Moreover, we have observed that it is beneficial for accuracy to formulate the (non)linear systems in terms of the updates `z = (y - u) / dt` so that a stage is `y = u + dt * z`. This is described in Hairer & Wanner II.